Feat/bancor v2

xianny opened the pull request:
## Description

Add integration for Bancor V2.

**contracts/asset-proxy**
* BancorBridge contract and tests

**asset-swapper**
* BancorService (wrapper for the Bancor SDK)
* tests for BancorService against mainnet and real Bancor network
* refactor of SourceQuoteOperation to return one fillData per sample instead of one fillData per source

**contract-addresses**
* added bancorBridge field

We're waiting on https://github.com/bancorprotocol/sdk/pull/20 to merge and publish before we enable Bancor as a source. After the SDK is updated, we need to do the following: 

- [ ] Deploy BancorBridge contract and then update @0x/contract-addresses
- [ ] Add Bancor to list of valid Sell sources in `.../market_operation_utils/constants.ts`
- [ ] Privately publish and AB test on staging/prod

## Testing instructions

I tested with simbot using https://github.com/dorothy-zbornak/0x-swap-server. Charts are in a comment below. 

Unit tests are updated.

## Types of changes
<!-- * Bug fix (non-breaking change which fixes an issue) -->

 * New feature (non-breaking change which adds functionality) 

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

xianny committed:
Bancor Bridge contract

xianny committed:
add sampling for bancor

xianny committed:
exclude instant from CI build

xianny committed:
add unit tests for bancor in dex sampler

dekz commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+    DeploymentConstants
+{
+
+    IContractRegistry contractRegistry = IContractRegistry(0x52Ae12ABe5D8BD778BD5397F99cA900624CfADD4);
```
we should have these addresses in DeploymentConstants so they can be faked in tests (with `address(this)`)

dekz commented on contracts/asset-proxy/contracts/src/interfaces/IBancorNetwork.sol:
```diff
@@ -0,0 +1,50 @@
+
+contract IBancorNetwork {
+    function convertByPath(
+        IERC20Token[] calldata _path,
```
replacing these with `address[]` then we can avoid `state.tokenPath[i] = IERC20Token(state.addressPath[i]);`

dekz commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+
+        // Otherwise use Bancor to convert
+        // Convert the addresses in the path to IERC20Tokens
+        for (uint i=0; i < state.addressPath.length; i++) {
```
we should be able to avoid this by  changing the interface of Bancor Contracts to be `address` or `address[]` instead of `IERC20Token`

dekz commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+            fromTokenBalance, // amount of source token to trade
+            amount, // minimum amount of destination token expected to receive
+            to, // beneficiary
+            address(0x0), // affiliateAccount; no fee paid
```
```suggestion
            address(0), // affiliateAccount; no fee paid
```

dekz commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+        LibERC20Token.approveIfBelow(state.addressPath[0], contractRegistry.addressOf(bancorNetworkName), fromTokenBalance);
+
+        // Convert the tokens
+        IBancorNetwork bancorNetwork = getBancorNetworkContract();
```
Worth putting this in the `bridgeData` to save the lookup, especially if it is essentially constant (e.g doesn't change depending on the tokens). Would save a chunk of gas.

dekz commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+
+        // // Grant an allowance to the Bancor Network to spend `fromTokenAddress` token.
+        uint256 fromTokenBalance = state.tokenPath[0].balanceOf(address(this));
+        LibERC20Token.approveIfBelow(state.addressPath[0], contractRegistry.addressOf(bancorNetworkName), fromTokenBalance);
```
same with comment as below

dekz commented on packages/asset-swapper/src/utils/market_operation_utils/bancor_service.ts:
```diff
@@ -0,0 +1,36 @@
+
+    public async getSDKAsync(): Promise<SDK> {
+        if (!this._sdk) {
+            this._sdk = await SDK.create({ ethereumNodeEndpoint: this.ethereumNodeEndpoint });
```
is there a way to give it a `provider` instead of having to pass down the RPC url?

xianny replied:
Haven't found one unfortunately :/

dorothy-zbornak replied:
It looks like you can pass in a web3 provider, might try just wrapping our provider in something like `{ sendAsync(payload) => provider.send(payload) }` or whatever works. https://github.com/bancorprotocol/sdk/blob/master/src/types.ts

dorothy-zbornak replied:
Does the above hack ^ work? Would also rather not pass in an RPC url.

xianny replied:
yeaaa it works! sorry I missed it on the first go round.

dekz commented on packages/asset-swapper/src/utils/market_operation_utils/bancor_service.ts:
```diff
@@ -0,0 +1,36 @@
+
+    public async getQuoteAsync(fromToken: string, toToken: string, amount: BigNumber): Promise<BancorQuoteData> {
+        const sdk = await this.getSDKAsync();
+        const { path, rate } = await sdk.pricing.getPathAndRate(token(fromToken), token(toToken), amount.toString());
```
does this make an RPC request for every amount? That seems like a lot of (13) requests?

dorothy-zbornak replied:
I *think* it actually caches the reserve graph and finds paths based off that. https://github.com/bancorprotocol/sdk/blob/453a13bf6d7d61e7136a004eabf6bc1f242b6f0e/src/blockchains/ethereum/index.ts#L77

This does seem to imply that we need a timer to routinely call `sdk.refresh()` though.

xianny replied:
yeah it does :/ paths are cached but each call to `getPathAndRate` refreshes the available rates in liquidity pools. basically every call to `getPathAndRate` results in 1 RPC call.

xianny replied:
chatted off-thread. we are going to leave it this way for now, pending bancor SDK upgrade to accept multiple amounts.

dekz commented on packages/asset-swapper/src/utils/market_operation_utils/bancor_service.ts:
```diff
@@ -0,0 +1,36 @@
+        const sdk = await this.getSDKAsync();
+        const { path, rate } = await sdk.pricing.getPathAndRate(token(fromToken), token(toToken), amount.toString());
+        return {
+            fillData: { path: path.map(p => p.blockchainId) }, // ethereum token addresses
```
this may also give different paths given different amounts?
E.g $100 gives A->B, $500 gives A->B->C. The sampler expects these to all be the same "route/source" so can encounter errors trying to fill for $600 if A->B or A->B->C do not have sufficient liquidity.

We should see if there are SDK functions that allow you to efficiently query with a hint of a path, or like Balancer, select a few candidates and query those.

dorothy-zbornak replied:
I wonder if we can just use the fillData from the last fill node when constructing the order.

xianny replied:
we can query with a full path or with source + target token, but I don't think we can provide a partial path. 

I'm making changes so `fillData.path` is unique to the amount quoted, so that should resolve the fill data problem. But perhaps still inefficient with the number of RPC calls per quote (1 per quote amount).

xianny replied:
did a deeeeep dive with @dorothy-zbornak in current implementation of path optimisation. as suggested, we can use the `fillData` from the last node.

dekz commented on packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:
```diff
@@ -287,6 +291,42 @@ export const samplerOperations = {
+            },
+            handleCallResultsAsync: async (_contract, _callResults) => {
+                return Promise.all(
+                    makerFillAmounts.map(amt => bancorService.getQuoteAsync(takerToken, makerToken, amt)), // tslint:disable-line:promise-function-async
```
Without a side this would be ambiguous and incorrect. 

Buying is `receive 100 of makerToken by selling X takerToken`. 
Selling is `receive X of makerToken by selling 100 takerToken`.

This is performing `receive X of takerToken by selling 100 makerToken`.

dorothy-zbornak replied:
Doesn't `getPathAndRate()` take a string for `amt`? https://github.com/bancorprotocol/sdk/blob/master/src/pricing.ts#L17

dorothy-zbornak replied:
Yeah let's just backlog buy support for bancor right now?

xianny replied:
made changes for `getSellQuote`, plz confirm sensical

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+            state.addressPath,
+            state.minReturn
+        // solhint-disable indent
+        ) = abi.decode(bridgeData, (address[], uint));
```
Should be able to just do `abi.decode(IERC20Token[], ...` so it fits with all the types

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,135 @@
+{
+
+    IContractRegistry contractRegistry = IContractRegistry(0x52Ae12ABe5D8BD778BD5397F99cA900624CfADD4);
+    bytes32 bancorNetworkName = bytes32(uint(0x42616e636f724e6574776f726b)); // "BancorNetwork"
```
```suggestion
    bytes32 internal constant bancorNetworkName = bytes32(uint(0x42616e636f724e6574776f726b)); // "BancorNetwork"
```

xianny committed:
Changes after review

- supply bancor network address in bridgeData
- write tests for bancor bridge
- write tests for bancor sdk
- skip Bancor buy quotes
- fix fillData in `collapsePath`

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -0,0 +1,122 @@
+        ) = abi.decode(bridgeData, (address[], address));
+        // solhint-enable indent
+
+        require(state.path.length > 2, "BancorBridge/PATH_LENGTH_MUST_BE_GREATER_THAN_TWO");
+        require(state.path[state.path.length - 1] == toTokenAddress, "BancorBridge/LAST_ELEMENT_OF_PATH_MUST_MATCH_OUTPUT_TOKEN");
+
```
So the shortest path is `[fromToken, converter, toToken]`, right?

xianny replied:
yup correct

dorothy-zbornak commented on contracts/asset-proxy/contracts/test/TestBancorBridge.sol:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    /// @dev Sets the balance of this contract for an existing token.
+    ///      The wei attached will be the balance.
```
```suggestion
```

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/index.ts:
```diff
@@ -241,21 +244,23 @@ export class MarketOperationUtils {
                         this._liquidityProviderRegistry !== NULL_ADDRESS ? [ERC20BridgeSource.LiquidityProvider] : [],
                     ),
-                    _opts.excludedSources.concat(ERC20BridgeSource.Balancer),
+                    _opts.excludedSources.concat([ERC20BridgeSource.Balancer]),
```
:see_no_evil:

xianny replied:
`concat` works on individual elements *and* arrays because types aren't real in node 😭 

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/bancor_service.ts:
```diff
@@ -0,0 +1,53 @@
+                path: path.map(p => p.blockchainId),
+                networkAddress: await this.getBancorNetworkAddressAsync(),
+            }, // ethereum token addresses
+            amount: new BigNumber(rate).multipliedBy(this.minReturnAmountBufferPercentage),
```
Do we need to round this to an integer?

xianny replied:
discussed off-thread. amounts have to be converted to/from wei instead.

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:
```diff
@@ -456,15 +485,35 @@ export const samplerOperations = {
+        const samplerOps: Array<SourceQuoteOperation<FillData, QuoteData>> = [];
+        const nonSamplerOps: Array<SourceQuoteOperation<FillData, QuoteData>> = [];
+        subOps.forEach(op => {
+            if (nonSamplerSources.find(s => s === op.source) !== undefined) {
```
```suggestion
            if (nonSamplerSources.includes(op.source)) {
```

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/types.ts:
```diff
@@ -265,7 +277,8 @@ export interface BatchedOperation<TResult> {
 }
 
-export interface SourceQuoteOperation<TFillData extends FillData = FillData> extends BatchedOperation<BigNumber[]> {
+export interface SourceQuoteOperation<TFillData extends FillData = FillData, TQuoteData extends QuoteData = BigNumber>
```
We should just make it so *all* sources return `BancorQuoteData` (but just called `QuoteData`) in `handleCallResultAsync()`. This one edge case makes the code confusing.

xianny replied:
refactored

xianny committed:
add bancor address

xianny committed:
Fixes

* accept wei in BancorService
* accept provider instead of rpc url

xianny committed:
refactor Quote and FillData types

xianny committed:
update changelogs

dorothy-zbornak commented on contracts/asset-proxy/contracts/src/bridges/BancorBridge.sol:
```diff
@@ -69,17 +69,16 @@ contract BancorBridge is
-        require(state.path[state.path.length - 1] == toTokenAddress, "BancorBridge/LAST_ELEMENT_OF_PATH_MUST_MATCH_OUTPUT_TOKEN");
-
         // Just transfer the tokens if they're the same.
         if (state.path[0] == toTokenAddress) {
```
~Will~ Can revert without the `path.length` check.

xianny replied:
If it's the same token, then `path.length = 1`. I'll add a `path.length > 0` check.

dorothy-zbornak commented on packages/asset-swapper/src/swap_quoter.ts:
```diff
@@ -201,7 +201,7 @@ export class SwapQuoter {
             },
         );
-        const bancorService = options.ethereumRpcUrl ? new BancorService(options.ethereumRpcUrl) : undefined;
+        const bancorService = options.ethereumRpcUrl ? new BancorService(provider) : undefined;
```
:tada:

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/bancor_service.ts:
```diff
@@ -33,15 +35,23 @@ export class BancorService {
+    ): Promise<Quote<BancorFillData>> {
         const sdk = await this.getSDKAsync();
-        const { path, rate } = await sdk.pricing.getPathAndRate(token(fromToken), token(toToken), amount.toString());
+        const blockchain = sdk._core.blockchains[BlockchainType.Ethereum] as Ethereum;
```
Wondering: can we set `sdk._core.blockchains[BlockchainType.Ethereum]` to a custom `Ethereum` instance that batches RPC calls?

dorothy-zbornak commented on contracts/asset-proxy/test/bancor_bridge.ts:
```diff
@@ -108,7 +108,10 @@ blockchainTests.resets('Bancor unit tests', env => {
                 // Transfer amount.
                 _opts.amount,
                 // ABI-encode the input token address as the bridge data. // FIXME
```
`FIXME`?

xianny replied:
woops that's outdated

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:
```diff
@@ -530,13 +561,9 @@ export const samplerOperations = {
                         input: takerFillAmounts[j],
-                        fillData:
-                            op.source === ERC20BridgeSource.Bancor ? (output as BancorQuoteData).fillData : op.fillData, // Bancor has unique fillData per sample
+                        fillData: output.fillData,
```
:+1:

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:
```diff
@@ -621,8 +644,8 @@ export const samplerOperations = {
+        const samplerOps: Array<SourceQuoteOperation<FillData>> = [];
+        const nonSamplerOps: Array<SourceQuoteOperation<FillData>> = [];
         subOps.forEach(op => {
             if (nonSamplerSources.find(s => s === op.source) !== undefined) {
```
```suggestion
            if (nonSamplerSources.includes(op.source)) {
```

dorothy-zbornak commented on packages/asset-swapper/src/utils/market_operation_utils/types.ts:
```diff
@@ -101,11 +101,9 @@ export interface BancorFillData extends FillData {
-export interface QuoteData {}
-
-export interface BancorQuoteData extends QuoteData {
-    fillData: BancorFillData;
```
:+1:

dorothy-zbornak reviewed (approved):
Just minor nits, but wtg!
![doublicat](https://user-images.githubusercontent.com/48803443/90686249-0f8af680-e239-11ea-8cf7-aed1618986f1.gif)

xianny committed:
minor nits

xianny committed:
Merge branch 'development' into feat/bancor-v2

xianny committed:
fixes after simbotting

xianny committed:
disable bancor while waiting for bancor SDK update

xianny committed:
add bancor to test

xianny commented:
Simbotted with Bancor enabled in https://github.com/dorothy-zbornak/0x-swap-server and got some data:

![Screen Shot 2020-08-21 at 9 29 11 AM](https://user-images.githubusercontent.com/8582774/90913327-f4cb9580-e390-11ea-81e8-dceb2cc4caf6.png)
No surprise here.



![Screen Shot 2020-08-21 at 9 28 12 AM](https://user-images.githubusercontent.com/8582774/90913349-feed9400-e390-11ea-8f76-43cd30449c69.png)
Bancor revert rate a bit worrying.


![Screen Shot 2020-08-21 at 9 28 45 AM](https://user-images.githubusercontent.com/8582774/90913503-33615000-e391-11ea-8cbe-aaad6ef28d28.png)
Bancor slippage also higher than the rest.



![Screen Shot 2020-08-21 at 9 30 03 AM](https://user-images.githubusercontent.com/8582774/90913525-3d834e80-e391-11ea-9a59-e5e91b976b6a.png)
Bancor TOKEN slippage is good though. Maybe we enable Bancor only for BNT for now??



![Screen Shot 2020-08-21 at 9 24 57 AM](https://user-images.githubusercontent.com/8582774/90913038-81298880-e390-11ea-9409-f4868ea9821d.png)
Just for the record.


---

A-B testing with prod:

![Screen Shot 2020-08-21 at 9 37 53 AM](https://user-images.githubusercontent.com/8582774/90914075-1c6f2d80-e392-11ea-8bf7-c7e7fe3959ba.png)
![Screen Shot 2020-08-21 at 9 38 25 AM](https://user-images.githubusercontent.com/8582774/90914084-1f6a1e00-e392-11ea-9270-8a7919e08160.png)
![Screen Shot 2020-08-21 at 9 36 01 AM](https://user-images.githubusercontent.com/8582774/90914104-27c25900-e392-11ea-9d4b-0d1ebfa21b60.png)
![Screen Shot 2020-08-21 at 9 37 03 AM](https://user-images.githubusercontent.com/8582774/90914124-30b32a80-e392-11ea-892c-948f308bd72a.png)
![Screen Shot 2020-08-21 at 9 36 36 AM](https://user-images.githubusercontent.com/8582774/90914179-49234500-e392-11ea-9691-9f2bae39f831.png)

xianny commented:
With latest asset-swapper (Bancor disabled)

![Screen Shot 2020-08-21 at 1 18 02 PM](https://user-images.githubusercontent.com/8582774/90931230-189ed380-e3b1-11ea-9b3d-10f33cc457f9.png)
![Screen Shot 2020-08-21 at 1 18 24 PM](https://user-images.githubusercontent.com/8582774/90931237-1b012d80-e3b1-11ea-8db7-ae905f66edd4.png)
![Screen Shot 2020-08-21 at 1 19 00 PM](https://user-images.githubusercontent.com/8582774/90931240-1ccaf100-e3b1-11ea-98d8-644ab7db37e2.png)

---
Little bit worried about Balancer reverts looking uneven so I ran it for a while with just Balancer and Uniswap as sources:
![Screen Shot 2020-08-21 at 3 06 26 PM](https://user-images.githubusercontent.com/8582774/90939272-155f1400-e3c0-11ea-8596-2dfe61c278d8.png)

---

![Screen Shot 2020-08-21 at 1 19 37 PM](https://user-images.githubusercontent.com/8582774/90931242-1f2d4b00-e3b1-11ea-90f8-f34228bba170.png)
![Screen Shot 2020-08-21 at 1 19 59 PM](https://user-images.githubusercontent.com/8582774/90931251-218fa500-e3b1-11ea-87d8-e7c903b5f3f0.png)
